pkg/validation-webhook/snapshot.go

```diff
@@ -0,0 +1,245 @@
+	"k8s.io/klog"
+)
+
+var (
+	// SnapshotGVR is GroupVersionResource for volumesnapshots
+	SnapshotGVR = metav1.GroupVersionResource{Group: snapshotGroup, Version: "v1beta1", Resource: "volumesnapshots"}
+	// SnapshotContentGVR is GroupVersionResource for volumesnapshotcontents
+	SnapshotContentGVR = metav1.GroupVersionResource{Group: snapshotGroup, Version: "v1beta1", Resource: "volumesnapshotcontents"}
+)
+
+// snapshot CRD api group
+const snapshotGroup = "snapshot.storage.k8s.io"
+
+// Add a label {"added-label": "yes"} to the object
+func admitSnapshot(ar v1.AdmissionReview) *v1.AdmissionResponse {
+	klog.V(2).Info("admitting volumesnapshots or volumesnapshotcontents")
+	fmt.Println("admitting volumesnapshots or volumesnapshotcontents")
+
+	reviewResponse := v1.AdmissionResponse{}
+	reviewResponse.Allowed = true
+
+	// Ignore Delete and Connect requests.
+	if !(ar.Request.Operation == v1.Update || ar.Request.Operation == v1.Create) {
+		return &reviewResponse
+	}
+	isUpdate := ar.Request.Operation == v1.Update
+
+	raw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	deserializer := codecs.UniversalDeserializer()
+	switch ar.Request.Resource {
+	case SnapshotGVR:
+		snapshot := &volumesnapshotv1beta1.VolumeSnapshot{}
+		if _, _, err := deserializer.Decode(raw, nil, snapshot); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSnapshot := &volumesnapshotv1beta1.VolumeSnapshot{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSnapshot); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		return decideSnapshot(snapshot, oldSnapshot, isUpdate)
+	case SnapshotContentGVR:
+		snapcontent := &volumesnapshotv1beta1.VolumeSnapshotContent{}
+		if _, _, err := deserializer.Decode(raw, nil, snapcontent); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSnapcontent := &volumesnapshotv1beta1.VolumeSnapshotContent{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSnapcontent); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		return decideSnapshotContent(snapcontent, oldSnapcontent, isUpdate)
+	default:
+		err := fmt.Errorf("expect resource to be %s or %s", SnapshotGVR, SnapshotContentGVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func decideSnapshot(snapshot, oldSnapshot *volumesnapshotv1beta1.VolumeSnapshot, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := v1.AdmissionResponse{}
+	reviewResponse.Allowed = true
+	reviewResponse.Result = &metav1.Status{}
+
+	// if it is an UPDATE and oldSnapshot is not valid, then don't enforce strict validation
+	// This allows no-op updates to occur on snapshot resources which fail strict validation
+	// Which allows the remover of finalizers and therefore deletion of this object
+	err := IsValidSnapshot(oldSnapshot)
+	if isUpdate && err != nil {
+		return &reviewResponse
+	}
+
+	err = IsValidSnapshot(snapshot)
```

Comment:
what if snapshot == nil? is it considered as a valid one or invalid one?

Reply:
Added a comment.

Don't rely on the pointers to be nil, because the deserialization method will convert it to the empty struct value. Instead check the operation type.

Reply:
The validation function will return an error if given a nil pointer, but i don't expect that to happen here. Instead we check if the operation is CREATE or UPDATE.